adapters/lora-vertx-base/src/main/java/org/eclipse/hono/adapter/lora/LoraProtocolAdapter.java

```diff
@@ -235,78 +238,106 @@ protected void customizeDownstreamMessageProperties(final Map<String, Object> pr
-                    handle202(ctx.getRoutingContext());
-                }
-            } catch (final LoraProviderMalformedPayloadException e) {
-                LOG.debug("error processing request from provider [name: {}]", provider.getProviderName(), e);
-                TracingHelper.logError(currentSpan, "error processing request", e);
-                handle400(ctx.getRoutingContext(), ERROR_MSG_INVALID_PAYLOAD);
+                Optional.ofNullable(uplinkMessage.getAdditionalData())
+                        .ifPresent(additionalData -> ctx.put(LoraConstants.APP_PROPERTY_ADDITIONAL_DATA, additionalData));
+
+                final String contentType = payload.length() > 0
+                        ? LoraConstants.CONTENT_TYPE_LORA_BASE + provider.getProviderName()
+                        : EventConstants.CONTENT_TYPE_EMPTY_NOTIFICATION;
+
+                currentSpan.finish(); // uploadTelemetryMessage will finish the root span, therefore finish child span here already
+                uploadTelemetryMessage(ctx, gatewayDevice.getTenantId(), deviceId, payload, contentType);
+                registerCommandConsumerIfNeeded(provider, gatewayDevice, currentSpan.context());
+                break;
+            default:
+                LOG.debug("discarding message of unsupported type [tenant: {}, device-id: {}, type: {}]",
+                        gatewayDevice.getTenantId(), deviceId, type);
+                currentSpan.log("discarding message of unsupported type");
+                currentSpan.finish();
+                // discard the message but return 202 to not cause errors on the LoRa provider side
+                handle202(ctx.getRoutingContext());
             }
-        } else {
-            handleUnsupportedUserType(ctx.getRoutingContext(), currentSpan);
+        } catch (final LoraProviderMalformedPayloadException e) {
+            LOG.debug("error processing request from provider [name: {}]", provider.getProviderName(), e);
+            TracingHelper.logError(currentSpan, "error processing request", e);
+            currentSpan.finish();
+            handle400(ctx.getRoutingContext(), ERROR_MSG_INVALID_PAYLOAD);
+        }
+    }
+
+    private void registerCommandConsumerIfNeeded(final LoraProvider provider, final Device gatewayDevice,
+            final SpanContext context) {
+        final String tenantId = gatewayDevice.getTenantId();
+        final String gatewayId = gatewayDevice.getDeviceId();
+        final SubscriptionKey key = new SubscriptionKey(tenantId, gatewayId);
+        if (commandSubscriptions.containsKey(key)) {
+            return;
         }
-        currentSpan.finish();
+        // use FOLLOWS_FROM span since this operation is decoupled from the rest of the request handling
+        final Span currentSpan = TracingHelper.buildFollowsFromSpan(tracer, context, "create command consumer")
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER)
```

Comment:
IMHO this span's kind should rather be *client* as we are now creating the consumer and invoke the Command Router service ...